module/zfs/dbuf.c

```diff
@@ -892,31 +906,46 @@ dbuf_init(void)
 	while (hsize * zfs_arc_average_blocksize < arc_all_memory() / 8)
 		hsize <<= 1;
 
-retry:
-	h->hash_table_mask = hsize - 1;
-#if defined(_KERNEL)
+	while (h->hash_table == NULL) {
+		h->hash_table_mask = hsize - 1;
+
+		h->hash_table = vmem_zalloc(hsize * sizeof (void *), KM_SLEEP);
+		if (h->hash_table == NULL)
+			hsize >>= 1;
+
+		ASSERT3U(hsize, >=, 1ULL << 10);
+	}
+
 	/*
-	 * Large allocations which do not require contiguous pages
-	 * should be using vmem_alloc() in the linux kernel
+	 * The hash table buckets are protected by an array of mutexes where
+	 * each mutex is reponsible for protecting 128 buckets.  On x86_64,
+	 * with 64 byte cachelines, this means the hash array will consume
+	 * 64KB per GB of total system memory.  A minimum value of 8192 is
+	 * targeted to maintain performance (512KB of memory).
 	 */
-	h->hash_table = vmem_zalloc(hsize * sizeof (void *), KM_SLEEP);
-#else
-	h->hash_table = kmem_zalloc(hsize * sizeof (void *), KM_NOSLEEP);
-#endif
-	if (h->hash_table == NULL) {
-		/* XXX - we should really return an error instead of assert */
-		ASSERT(hsize > (1ULL << 10));
-		hsize >>= 1;
-		goto retry;
+	if (dbuf_mutex_hash_shift == 0)
+		hmsize = MAX(hsize >> 7, 1ULL < 11);
+	else
+		hmsize = 1ULL << MIN(dbuf_mutex_hash_shift, 24);
+
+	while (h->hash_mutexes == NULL) {
+		h->hash_mutex_mask = hmsize - 1;
+
+		h->hash_mutexes = vmem_zalloc(hmsize *
```

Comment:
I do not see anything to align the memory here. There is no guarantee that `vmem_zalloc()` will align on cacheline boundaries. Your comment "You're mileage will vary by architecture" make me think that you had handled that.

This is a regression without alignment. To align on 64B cachelines, you want to overallocate by 64 bytes and then use P2ROUNDUP() to round up to the nearest 64B boundary. The rounded pointer would then go into `h->hash_mutexes`, while the original pointer needs to be kept around for module unloading.

Reply:
I am not 100% sure about Linux (but suspect also), but on FreeBSD all big memory allocations are aligned to page size and all smaller -- to the next power of 2, so no additional alignment should be needed here, assuming the array element size is a multiple of a cache size (on FreeBSD struct sx is 32 bytes, so yes).

Reply:
This is definitely the case for the Linux kernel, large allocations like this will be page-aligned.

Reply:
In that case, we are fine here.